stix_shifter_modules/nozomi/stix_transmission/connector.py

```diff
@@ -0,0 +1,216 @@
+                            next_url = None
+                            break
+                        # parsing the next_url
+                        page = regex.findall(r'page=([^>]*?)&', query_url)[0]
+                        next_url = regex.sub(r'page=([^>]*?)&', f'page={str(int(page) + 1)}&', query_url)
+
+                        if not metadata and result_count < total_records:
+                            remaining_records = total_records - result_count
+                        elif metadata and local_result_count < total_records:
+                            remaining_records = total_records - local_result_count
+                        else:
+                            break
+
+                        if remaining_records > Connector.NOZOMI_MAX_PAGE_SIZE:
+                            page_size = Connector.NOZOMI_MAX_PAGE_SIZE
+                        else:
+                            page_size = remaining_records
+                    # Handling expired token
+                    elif response_wrapper.code == 401 and not token_generated and \
+                            (response == '' or
+                             'Signature has expired' in response_dict.get('error', {}).get('message', '')):
+                        jwt_token = await self.__get_token()
+                        token_generated = True
+                        continue
+                    else:
+                        error = response_dict.get('error', '')
+                        if response_wrapper.code == 403:
+                            error = 'Query length is too long or Invalid Query'
+                        return_obj = self.handle_api_exception(response_wrapper.code, error)
+                        data = []
+                        break
+
+                if data:
+                    data = self.get_results_data(data)
+                    if metadata:
+                        return_obj['data'] = data if data else []
+                    else:
+                        return_obj['data'] = data[int(offset): total_records] if data else []
+                    if jwt_token and next_url and result_count < self.api_client.result_limit:
+                        return_obj['metadata'] = {"result_count": result_count, "jwtToken": jwt_token,
+                                                  "next_url": next_url}
+                else:
+                    if not return_obj.get('error') and return_obj.get('success') is not False:
+                        return_obj['success'] = True
+                        return_obj['data'] = []
+            else:
+                return_obj['success'] = True
+                return_obj['data'] = []
+
+        except InvalidMetadataException as ex:
+            return_obj = self.handle_api_exception(422, f'Invalid metadata: {str(ex)}')
+
+        except Exception as ex:
+            return_obj = self.handle_api_exception(None, str(ex))
+        return return_obj
+
```

Comment:
I have a few general comments for how this method works and it's probably easier to understand with a bigger post explaining the intricacies here.

The first issue I have is that the page size seems to be variable based off the length of the request. While this works, conceptually it's strange to vary your page size. If you reach the event limit in a single page, it should simply grab the (X) newest or oldest events and return that data. The user should be able to modify the page size they want to use, but once it's set, it's final (unless they decide to start a new query). It should also be limited to 1000 or less. If there are less results than the page size, that's not a big deal, it just means you got all of the events.

The concept is that the query should really not change much between calls, the only difference would be the current page which starts at 1 or at metadata[current_page]. You wouldn't need to worry about regex as you know the current page at all times and you know the next page is just the current_page+1. This should remove the somewhat unnecessary regex parsing for pages. When the page is partially completed but the total_events was reached, then you simply keep the current_page in the meta_data for the next call. 

To handle partial data (you reached the data_limit mid-page), you should return a metadata[page_indice] that indicates what the indice of the last completed page was. Thus when you continue the next call, it uses that page_indice and collects any remaining data from the page that was not previously completed.

Now there is an additional problem with the page limit being 1,000. This limit means that if you do reach the maximum page count, you need to construct a new query to capture the remaining data. Their documentation suggest
"it is advisable to set up a time field pivot when reaching page 1000 and start again from page 1.". I'd suggest ensuring that all of the queries are sorted based off the time field and when you reach the page 1000 limit, you simply update the minimum time in the query to match that the newest time in the data set (I'd recommend oldest -> newest, but newest -> oldest can work too).

Reply:
Below is the list of changes updated in transmit part.
	1. As per the recommendation followed max page size is 1000.
	2. Removed next_url and jwt_token in metadata. As per the suggestion using page_size and page_number in metadata.
	3. Removed regex parsing.
	4. If page_number is more than 1000 it will throw the error.